pkg/reconciler/buildrun/resources/build.go

```diff
@@ -8,27 +8,43 @@ import (
 	"context"
 	"fmt"
 
-	buildv1alpha1 "github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	buildv1alpha1 "github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
 )
 
 // GetBuildObject retrieves an existing Build based on a name and namespace
 func GetBuildObject(ctx context.Context, client client.Client, buildRun *buildv1alpha1.BuildRun, build *buildv1alpha1.Build) error {
-	err := client.Get(ctx, types.NamespacedName{Name: buildRun.Spec.BuildRef.Name, Namespace: buildRun.Namespace}, build)
-	if err != nil {
+	// Option #1: BuildRef is specified
+	// An actual Build resource is specified by name and needs to be looked up in the cluster.
+	if buildRun.Spec.BuildRef != nil {
+		err := client.Get(ctx, types.NamespacedName{Name: buildRun.Spec.BuildName(), Namespace: buildRun.Namespace}, build)
 		if apierrors.IsNotFound(err) {
 			// stop reconciling and mark the BuildRun as Failed
 			// we only reconcile again if the status.Update call fails
-			if updateErr := UpdateConditionWithFalseStatus(ctx, client, buildRun, fmt.Sprintf("build.shipwright.io \"%s\" not found", buildRun.Spec.BuildRef.Name), ConditionBuildNotFound); updateErr != nil {
+			if updateErr := UpdateConditionWithFalseStatus(ctx, client, buildRun, fmt.Sprintf("build.shipwright.io %q not found", buildRun.Spec.BuildName()), ConditionBuildNotFound); updateErr != nil {
 				return HandleError("build object not found", err, updateErr)
 			}
 		}
+
+		return err
+	}
+
+	// Option #2: BuildSpec is specified
+	// The build specification is embedded in the BuildRun itself, create a transient Build resource.
+	if buildRun.Spec.BuildSpec != nil {
+		build.Name = "embedded-buildspec-build"
+		build.Namespace = buildRun.Namespace
+		build.Spec = *buildRun.Spec.BuildSpec
+		build.Status = buildv1alpha1.BuildStatus{}
+		return nil
 	}
 
-	return err
+	// Bail out hard in case of an invalid state (unreachable code)
+	panic("invalid BuildRun resource that neither has a build reference nor an embedded build specification")
```

Comment:
I don't think that a "getter" function should be unsafe and possibly crash the entire controller. Since the api of the function already offers to return an error, I would like to see an error for the panic case that is handled higher up.